### PR TITLE
rgw: rework json/xml escape usage follow #19806

### PR DESCRIPTION
**Pull request by fangyuxiangGL**
@cbodley 

rework follow your pr.

Signed-off-by: fang yuxiang fang.yuxiang@eisoo.com

### CODE REVIEW

File: src/cls/rgw/cls_rgw.cc [hunk 1]
```diff
@@ -989,10 +989,9 @@ static void update_olh_log(struct rgw_bucket_olh_entry& olh_data_entry, OLHLogOp
 
 static string escape_str(const string& s)
 {
-   int len = escape_json_attr_len(s.c_str(), s.size());
-   char escaped[len];
-   escape_json_attr(s.c_str(), s.size(), escaped);
-   return string(escaped);
+  std::stringstream ss;
+  ss << json_stream_escaper(s.c_str());
+  return ss.str();
```
[the remark this answers is not in the document]
Comment (cbodley) [reply src/cls/rgw/cls_rgw.cc]: hey neat, https://github.com/ceph/ceph/pull/19657 merged and we can use c++17 now :)

File: src/common/Formatter.cc [hunk 2]
```diff
@@ -163,10 +163,7 @@ void JSONFormatter::print_comma(json_formatter_stack_entry_d& entry)
 
 void JSONFormatter::print_quoted_string(const std::string& s)
 {
-  int len = escape_json_attr_len(s.c_str(), s.size());
-  char escaped[len];
-  escape_json_attr(s.c_str(), s.size(), escaped);
-  m_ss << '\"' << escaped << '\"';
+  m_ss << '\"' << json_stream_escaper(s.c_str()) << '\"';
```
[the remark this answers is not in the document]
Comment (fangyuxiangGL) [reply src/common/Formatter.cc]: yeah, i ignored the efficiency when making the changes..

File: src/cls/rgw/cls_rgw.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (fangyuxiangGL) [reply src/cls/rgw/cls_rgw.cc]: @cbodley 

#19657 was already merged, and I have rebased to it.

File: src/common/Formatter.cc [hunk 3]
```diff
@@ -566,10 +563,9 @@ void XMLFormatter::print_spaces()
 
 std::string XMLFormatter::escape_xml_str(const char *str)
 {
-  int len = escape_xml_attr_len(str);
-  std::vector<char> escaped(len, '\0');
-  escape_xml_attr(str, &escaped[0]);
-  return std::string(&escaped[0]);
+  std::stringstream ss;
+  ss << xml_stream_escaper(str);
+  return ss.str();
```
[the remark this answers is not in the document]
Comment (fangyuxiangGL) [reply src/common/Formatter.cc]: yeah, already addressed it. thx

File: src/common/escape.h [hunk 4]
```diff
@@ -18,28 +18,6 @@
...
- * as an JSON attrribute
- */
-int escape_json_attr_len(const char *buf, int src_len);
-
-/* Escapes 'buf' as an JSON attribute. Assumes that 'out' is at least long
- * enough to fit the output. You can find out the required length by calling
- * escape_json_attr_len first.
- */
-void escape_json_attr(const char *buf, int src_len, char *out);
-
```
[the remark this answers is not in the document]
Comment (fangyuxiangGL) [reply src/common/escape.h]: make sense, it's reverted.

**Comment by fangyuxiangGL** [comment]
@cbodley 

updates.

**Comment by fangyuxiangGL** [comment]
jenkis retest plz

File: src/rgw/rgw_formats.cc [hunk 5]
```diff
@@ -287,7 +287,9 @@ void RGWFormatter_Plain::dump_value_int(const char *name, const char *fmt, ...)
 class HTMLHelper : public XMLFormatter {
 public:
   static std::string escape(const std::string& unescaped_str) {
-    return escape_xml_str(unescaped_str.c_str());
+    std::stringstream ss;
+    ss << xml_stream_escaper(unescaped_str);
+    return ss.str();
```
Comment (cbodley) [inline src/rgw/rgw_formats.cc]: a few of the callers of `HTMLHelper::escape()` are using streams, so you can use `xml_stream_escaper` directly in those cases. the rest are being passed into `boost::format`, so we do need them as strings - so lets use the same approach here with `std::string::data()` as you did for `escape_str()` in rgw_cls.cc

**Comment by fangyuxiangGL** [comment]
@cbodley 

updates. could you have a look if there is anywhere leaked?
thanks.

**Comment by yuriw** [comment]
[wip-yuri3-testing-2018-01-10-1647](https://trello.com/c/7GyGI8yl/226-wip-yuri3-testing-2018-01-10-1647)

**Comment by yuriw** [comment]
suspecting errors http://pulpito.ceph.com/yuriw-2018-01-12_23:43:43-rgw-wip-yuri3-testing-2018-01-12-1646-distro-basic-smithi/ untagged for now

**Comment by fangyuxiangGL** [comment]
@cbodley @yuriw 

I took a look at the error log but I couldn't find useful information, could you give some help? thx

**Comment by yuriw** [comment]
[wip-yuri2-testing-2018-01-16-1725](https://trello.com/c/PBlOeZJA/229-wip-yuri2-testing-2018-01-16-1725)

File: src/rgw/rgw_formats.cc [hunk 6]
```diff
@@ -287,7 +287,10 @@ void RGWFormatter_Plain::dump_value_int(const char *name, const char *fmt, ...)
 class HTMLHelper : public XMLFormatter {
 public:
   static std::string escape(const std::string& unescaped_str) {
-    return escape_xml_str(unescaped_str.c_str());
+    int len = escape_json_attr_len(unescaped_str.c_str(), unescaped_str.size());
+    std::string escaped(len, 0);
+    escape_json_attr(unescaped_str.c_str(), unescaped_str.size(), escaped.data());
+    return escaped;
```
Comment (cbodley) [inline src/rgw/rgw_formats.cc]: this should be calling the xml functions, not json

**Comment by cbodley** [comment]
@fangyuxiangGL i dug into the first `s3tests.functional.test_s3.test_object_copy_zero_size` failure, and see that `RGWCopyObj_ObjStore_S3::send_response()` is dumping this:
```
<LastModified>2018-01-16T19:12:01.000Z</LastModified><ETag>d41d8cd98f00b204e9800998ecf8427e&#x00;</ETag></CopyObjectResult>
```
where before, it was:
```
<LastModified>2018-01-16T19:50:39.453Z</LastModified><ETag>d41d8cd98f00b204e9800998ecf8427e</ETag></CopyObjectResult>
```

so the issue here is how we handle the null termination on the etag (which is coming from a null-terminated bufferlist, so `bufferlist::to_str()` ends up with an extra null at the end). i see that `escape_xml_attr()` has a `for (b = buf; *b; ++b)` loop that stops on the null terminator, but the new stream escapers rely on the string_view's size instead. the existing loop in `escape_json_attr()` works the same based on its `src_len` parameter

i think the `stream_escapers` are doing the right thing here, so i'd recommend stripping the null off of the etag there, like:
```patch
--- a/src/rgw/rgw_rest_s3.cc
+++ b/src/rgw/rgw_rest_s3.cc
@@ -2153,7 +2153,7 @@ void RGWCopyObj_ObjStore_S3::send_response()
 
   if (op_ret == 0) {
     dump_time(s, "LastModified", &mtime);
-    std::string etag_str = etag.to_str();
+    std::string etag_str = etag.c_str();
     if (! etag_str.empty()) {
       s->formatter->dump_string("ETag", std::move(etag_str));
     }
```
with that change applied, this branch passes our swift and s3tests

File: src/rgw/rgw_formats.cc (hunk 6 above)
Comment (fangyuxiangGL) [reply src/rgw/rgw_formats.cc]: Must be a unforgivable mistake, thanks!

**Comment by fangyuxiangGL** [comment]
@cbodley 

thanks for detailed explanation, i will also try s3test to verify it after apply the changes!

**Comment by fangyuxiangGL** [comment]
jenkins retest please

**Comment by fangyuxiangGL** [comment]
jenkins retest please

**Comment by fangyuxiangGL** [comment]
jenkins retest please

**Comment by fangyuxiangGL** [comment]
@cbodley Should we request qa to retest again? thanks

**Comment by fangyuxiangGL** [comment]
@yuriw 
Shall we have any updates? thanks

**Comment by cbodley** [comment]
@fangyuxiangGL we're starting up test runs again, now that luminous backport testing is done - could you please rebase this one?

**Comment by fangyuxiangGL** [comment]
@cbodley thanks, already rebased this.

**Comment by yuriw** [comment]
[wip-yuri4-testing-2018-02-26-1600](https://trello.com/c/AuozblD1/269-wip-yuri4-testing-2018-02-26-1600)

**Pull request merged**